File: packages/common/pydantic/product_rollups.py
"""Shared Pydantic contract for product-facing rollup presentations."""

from __future__ import annotations

from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field, model_validator

from .neuro_score_taxonomy import NeuroScoreStatus
from .readout_aggregate_metrics import SyntheticLiftCalibrationStatus


class ProductRollupMode(str, Enum):
    creator = "creator"
    enterprise = "enterprise"


class ProductRollupWarningSeverity(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"


class ProductLiftTruthStatus(str, Enum):
    unavailable = "unavailable"
    pending = "pending"
    measured = "measured"


class ProductScoreSummary(BaseModel):
    metric_key: str = Field(min_length=1, max_length=64)
    display_label: str = Field(min_length=1)
    scalar_value: Optional[float] = Field(default=None, ge=0, le=100)
    confidence: Optional[float] = Field(default=None, ge=0, le=1)
    status: NeuroScoreStatus
    explanation: str = Field(min_length=1)
    source_metrics: List[str] = Field(default_factory=list)


class ProductRollupWarning(BaseModel):
    warning_key: str = Field(min_length=1, max_length=128)
    severity: ProductRollupWarningSeverity
    message: str = Field(min_length=1)
    source_metrics: List[str] = Field(default_factory=list)


class CreatorProductRollups(BaseModel):
    reception_score: ProductScoreSummary
    organic_reach_prior: ProductScoreSummary
    explanations: List[str] = Field(default_factory=list)
    warnings: List[ProductRollupWarning] = Field(default_factory=list)


class ProductLiftComparison(BaseModel):
    synthetic_lift_prior: ProductScoreSummary
    predicted_incremental_lift_pct: Optional[float] = None
    predicted_iroas: Optional[float] = None
    predicted_incremental_lift_ci_low: Optional[float] = None
    predicted_incremental_lift_ci_high: Optional[float] = None
    measured_lift_status: ProductLiftTruthStatus = ProductLiftTruthStatus.unavailable
    measured_incremental_lift_pct: Optional[float] = None
    measured_iroas: Optional[float] = None
    calibration_status: Optional[SyntheticLiftCalibrationStatus] = None
    note: str = Field(min_length=1)


class EnterpriseDecisionSupport(BaseModel):
    media_team_summary: str = Field(min_length=1)
    creative_team_summary: str = Field(min_length=1)


class EnterpriseProductRollups(BaseModel):
    paid_lift_prior: ProductScoreSummary
    brand_memory_prior: ProductScoreSummary
    cta_reception_score: ProductScoreSummary
    synthetic_lift_prior: ProductScoreSummary
    synthetic_vs_measured_lift: ProductLiftComparison
    decision_support: EnterpriseDecisionSupport


class ProductRollupPresentation(BaseModel):
    mode: ProductRollupMode
    workspace_tier: str = Field(min_length=1, max_length=64)
    enabled_modes: List[ProductRollupMode] = Field(default_factory=list)
    mode_resolution_note: Optional[str] = None
    source_schema_version: str = Field(min_length=1)
    creator: Optional[CreatorProductRollups] = None
    enterprise: Optional[EnterpriseProductRollups] = None

    @model_validator(mode="after")
    def validate_mode_payload(self) -> "ProductRollupPresentation":
        if self.mode not in self.enabled_modes:
            raise ValueError("mode must be present in enabled_modes")
        if self.mode == ProductRollupMode.creator and self.creator is None:
            raise ValueError("creator payload must be provided when mode=creator")
        if self.mode == ProductRollupMode.enterprise and self.enterprise is None:
            raise ValueError("enterprise payload must be provided when mode=enterprise")
        return self


for _model in (
    ProductScoreSummary,
    ProductLiftComparison,
    CreatorProductRollups,
    EnterpriseProductRollups,
    ProductRollupPresentation,
):
    _model.model_rebuild()

File: packages/common/zod/productRollups.ts
import { z } from 'zod';
import { neuroScoreStatusSchema } from './neuroScoreTaxonomy';

export const productRollupModeSchema = z.enum(['creator', 'enterprise']);
export const productRollupWarningSeveritySchema = z.enum(['low', 'medium', 'high']);
export const productLiftTruthStatusSchema = z.enum(['unavailable', 'pending', 'measured']);
export const syntheticLiftCalibrationStatusSchema = z.enum([
  'uncalibrated',
  'provisional',
  'geox_calibrated',
  'truth_layer_unavailable'
]);

export const productScoreSummarySchema = z.object({
  metric_key: z.string().min(1).max(64),
  display_label: z.string().min(1),
  scalar_value: z.number().min(0).max(100).nullable(),
  confidence: z.number().min(0).max(1).nullable(),
  status: neuroScoreStatusSchema,
  explanation: z.string().min(1),
  source_metrics: z.array(z.string().min(1)).default([])
});

export const productRollupWarningSchema = z.object({
  warning_key: z.string().min(1).max(128),
  severity: productRollupWarningSeveritySchema,
  message: z.string().min(1),
  source_metrics: z.array(z.string().min(1)).default([])
});

export const creatorProductRollupsSchema = z.object({
  reception_score: productScoreSummarySchema,
  organic_reach_prior: productScoreSummarySchema,
  explanations: z.array(z.string().min(1)).default([]),
  warnings: z.array(productRollupWarningSchema).default([])
});

export const productLiftComparisonSchema = z.object({
  synthetic_lift_prior: productScoreSummarySchema,
  predicted_incremental_lift_pct: z.number().nullable().optional(),
  predicted_iroas: z.number().nullable().optional(),
  predicted_incremental_lift_ci_low: z.number().nullable().optional(),
  predicted_incremental_lift_ci_high: z.number().nullable().optional(),
  measured_lift_status: productLiftTruthStatusSchema.default('unavailable'),
  measured_incremental_lift_pct: z.number().nullable().optional(),
  measured_iroas: z.number().nullable().optional(),
  calibration_status: syntheticLiftCalibrationStatusSchema.nullable().optional(),
  note: z.string().min(1)
});

export const enterpriseDecisionSupportSchema = z.object({
  media_team_summary: z.string().min(1),
  creative_team_summary: z.string().min(1)
});

export const enterpriseProductRollupsSchema = z.object({
  paid_lift_prior: productScoreSummarySchema,
  brand_memory_prior: productScoreSummarySchema,
  cta_reception_score: productScoreSummarySchema,
  synthetic_lift_prior: productScoreSummarySchema,
  synthetic_vs_measured_lift: productLiftComparisonSchema,
  decision_support: enterpriseDecisionSupportSchema
});

export const productRollupPresentationSchema = z
  .object({
    mode: productRollupModeSchema,
    workspace_tier: z.string().min(1).max(64),
    enabled_modes: z.array(productRollupModeSchema).default([]),
    mode_resolution_note: z.string().nullable().optional(),
    source_schema_version: z.string().min(1),
    creator: creatorProductRollupsSchema.nullable().optional(),
    enterprise: enterpriseProductRollupsSchema.nullable().optional()
  })
  .superRefine((value, ctx) => {
    if (!value.enabled_modes.includes(value.mode)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['enabled_modes'],
        message: 'mode must be present in enabled_modes.'
      });
    }
    if (value.mode === 'creator' && !value.creator) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['creator'],
        message: 'creator payload must be provided when mode=creator.'
      });
    }
    if (value.mode === 'enterprise' && !value.enterprise) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['enterprise'],
        message: 'enterprise payload must be provided when mode=enterprise.'
      });
    }
  });

export type ProductRollupPresentation = z.infer<typeof productRollupPresentationSchema>;
